Add Usage to docs

phackstock opened the pull request:
Mostly self explanatory, added a "Usage" section to the docs which includes all parts of the documentation that did not fit into the index or the getting started page.

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+Usage
+=====
+
+Overview
+--------
+
+This sections dives deeper into the inner workings of **DataStructureDefinition** and
+**RegionProcessor** as well as their associated auxiliary classes.
+
+
+Yaml file schemas
+--------------------
+
+Both classes read information from yaml files which are required to follow specific
+structures.
```
Suggestion for a more concise, less verbose introduction.

```suggestion
The **DataStructureDefinition** and **RegionProcessor** classes
are initialized from yaml files that must follow specific formats.

This page describes the required specifications of the files.
```

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+Both classes read information from yaml files which are required to follow specific
+structures.
+
+For **DataStructureDefinition** there are three types of schemas: variable, region and tag.
```
Suggest to add a TOC at the top of this document and use the class names as section heading

phackstock replied:
Added a toc

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+* Other attributes specified in the yaml file can include (for countries)
+  ISO2/3-codes, or the list of countries included in a macro-region (i.e.,
+  a continent or large region).
+
```
"Generic" format should also be mentioned. This can be used, for example, to specify codelists for the allowed scenario names.

phackstock replied:
Added generic to the list

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+``<Tag>`` will also be replaced in any of the variable attributes.
+
+
+Mapping
+~~~~~~~
```
```suggestion
RegionProcessor
-----------------
```

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+
+-  The names of the three top level keywords **model**,
+   **native_regions** and **common_regions** are fixed.
+-  Required properties are **model** and **at least** either
```
Use different formatting from the fixed keywords
```suggestion
-  Required properties are **model** and *at least* either
```

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+-  Do not include words like *Level* or *Quantity* in the variable,
+   because this should be clear from the context or unit
+
+Class Overview
```
This section doesn't really make sense at the bottom. Best to integrate the descriptions into the text above.

phackstock replied:
Will do, should I start using the linked references for every mention of a class like you started in #77?
So :class:`nomenclature.codelist.Codelist` instead of **CodeLists**?

danielhuppmann replied:
Yes, we should use class-references wherever possible (that's why I started #77 today, to allow adopting this consistently right away).

phackstock replied:
Done

danielhuppmann commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,194 @@
+      the above example *region_a* and *region_b* and **not**
+      *alternative_name_a*.
+
+Guidelines and variable naming conventions
```
Suggest to move this to an own page "Guidelines"

phackstock replied:
Added a guidelines page

danielhuppmann reviewed (commented):
Thanks @phackstock - a first round of suggestions inline...

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+-  Do not use spaces before and after the *|* character, but add a
+   space between words (e.g., *Primary Energy|Non-Biomass Renewables*)
+-  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.)
+-  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary
```
```suggestion
-  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary.
-  Do not use abbreviations of statistical operations (*min*, *max*, *avg*) - always spell out the word.
```

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+
+The variable name (code) should adhere to the following conventions:
+
+-  A *|* (pipe) character indicates levels of hierarchy
```
```suggestion
-  A *|* (pipe) character indicates levels of hierarchy.
```

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+
+-  A *|* (pipe) character indicates levels of hierarchy
+-  Do not use spaces before and after the *|* character, but add a
+   space between words (e.g., *Primary Energy|Non-Biomass Renewables*)
```
```suggestion
   space between words (e.g., *Primary Energy|Non-Biomass Renewables*).
```

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+-  A *|* (pipe) character indicates levels of hierarchy
+-  Do not use spaces before and after the *|* character, but add a
+   space between words (e.g., *Primary Energy|Non-Biomass Renewables*)
+-  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.)
```
```suggestion
-  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.).
```

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+-  Add hierarchy levels where it might be useful in the future, e.g.,
+   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
+   Electric Vehicle*
+-  Do not use abbreviations of statistical operations (*min*, *max*,
```
```suggestion
```

phackstock replied:
Do you mean to delete this sentence or to move it up as well?

LauWien replied:
Yes, I'd move it up as well, like suggested in another suggestion - under the other comment on abbreviation maybe.

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
+   Electric Vehicle*
+-  Do not use abbreviations of statistical operations (*min*, *max*,
+   *avg*) but always spell out the word
```
```suggestion
```

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+Class Overview
+--------------
+
+A **DataStructureDefinition** class contains **CodeLists** for
```
**CodeLists** was written above small, we could either also use "codelists" here or above **CodeLists**. Or maybe I'm a bit confused ;)

phackstock replied:
I removed this section as it was no longer necessary.

LauWien commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+-  Do not include words like *Level* or *Quantity* in the variable,
+   because this should be clear from the context or unit
+
+Class Overview
```
Fits best in API documentation, since "...(a full list of all classes can be found in API documentation)...."

phackstock replied:
Removed this part as it was indeed not very useful here.

LauWien reviewed (commented):
Looks great! Some suggestions inline.

phackstock commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+The **DataStructureDefinition** and **RegionProcessor** classes
+are initialized from yaml files that must follow specific formats.
+
+This page describes the required specifications of the files.
```
Why should this be removed?

LauWien replied:
In another comment I moved it up.

phackstock replied:
Ah ok, since the your previous comment referred to an outdated version of the code I didn't see it in the files tab.

phackstock commented:
@LauWien I think we hit some bad timing yesterday since I pushed quite a lot of updates just before you did your review on the old version of the code. If it's ok for you I'll close all your comments that are outdated and request a re-review from you as soon as I'm ready to avoid further mix ups.

phackstock commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+     - Some Key:
+         description: a short description of the key
+
+When importing the codelist, any occurrence of ``<Tag>`` in a variable
```
Good question, I don't think so, worded like this it makes it sound like there can only be one tag codelist which is not the case.
I'll change it to something like:
When importing a tag codelist, any occurrence of <Tag> in a variable

phackstock commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+
+* Every region **must be** defined as part of a hierarchy. 
+
+* When importing the codelist, the hierarchy will be added as attribute,
```
No this only applies to the *region* codelist.

LauWien commented:
> @LauWien I think we hit some bad timing yesterday since I pushed quite a lot of updates just before you did your review on the old version of the code. If it's ok for you I'll close all your comments that are outdated and request a re-review from you as soon as I'm ready to avoid further mix ups.

Oh shoot, sorry, yes sounds good!

phackstock commented:
No problem it was just unfortunate timing, took me a minute as well to understand.

phackstock commented on doc/source/usage.rst:
```diff
@@ -0,0 +1,186 @@
+   DataStructureDefinition.region["Region Name"]["Hierarchy"] = "<Hierarchy Level>"
+
+* Other attributes specified in the yaml file can include (for countries)
+  ISO2/3-codes, or the list of countries included in a macro-region (i.e.,
```
Ah yes I'll put a wikipedia link

phackstock commented:
Another round of updates:
* Added a "generic" section to "usage"
* Added a special keyword section to "usage.variable"
* Fixed the references to code in nomenclature and pyam. 
* Moved the "guidelines" to its own page.
* Removed the "class overview" and spread the information throughout the document

phackstock commented:
@danielhuppmann do you think I should add another section to "usage" giving some details on the `process` function? I am mostly thinking about information on how to validate not all dimensions in `DataStructureDefinition`.

danielhuppmann commented:
> do you think I should add another section to "usage" giving some details on the process function?

I would not do that as part of this PR.

LauWien commented on doc/source/guidelines.rst:
```diff
@@ -0,0 +1,19 @@
+*  Do not use abbreviations of statistical operations (*min*, *max*,
+   *avg*) but always spell out the word.
+*  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.).
+*  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary.
```
I'd move this line up to line 13, so both abbreviations conventions are in sequence.

phackstock replied:
done

LauWien reviewed (approved):
LGTM :)

phackstock commented:
@danielhuppmann can I go ahead with the merge or do you want to review before?

danielhuppmann reviewed (approved):
Looks great, thanks!
